### ISSUE TITLE
Hang when hovering over constant

### ISSUE THREAD

**Issue by DaniPopes** [rust-lang/rust-analyzer#16907]
**rust-analyzer version**:
- 0.3.1885-standalone (b6d1887bc 2024-03-17)
- 0.4.1889-standalone (5e276ae51 2024-03-20)
- 0.0.0 (8d74705b43 2024-03-21)

**rustc version**: rustc 1.79.0-nightly (a7e4de13c 2024-03-19)

**relevant settings**: N/A

**repository link (if public, optional)**: N/A

**code snippet to reproduce**: 
```rust
pub enum AA {
    BB,
}

pub fn hang() -> AA {
    const CONST: AA = AA::BB;
    CONST
}
```

Hovering over either `CONST` starts the hang and rust-analyzer becomes unresponsive, consuming about ~200MiB RAM per second in this case, though in other cases it can be much faster.

The server cannot be killed through hovering on the `rust-analyzer` button on the lower bar or with `rust-analyzer: Stop server`, and must be killed through the command line. It is either non-responsive or doesn't kill the process.

I can reproduce this consistently with recent rust-analyzer client and server, but only after updating `rustc` from `nightly-2024-03-19` (`rustc 1.79.0-nightly (3c85e5624 2024-03-18)`) to `nightly-2024-03-20` (`rustc 1.79.0-nightly (a7e4de13c 2024-03-19)`) or later with rustup (override works too).

**Comment by Veykril** [comment]
https://github.com/rust-lang/rust/pull/122037 might be at fault here. Presumably this broke something in our mir eval

**Comment by HKalbasi** [comment]
https://github.com/rust-lang/rust/pull/122037 indeed needs some action, we need to consider function annotated with `#[rust_intrinsic]` or `#[rust_safe_intrinsic]` as intrinsic in addition to `extern "rust-intrinsic"` functions. But it is not clear to me how this simple example needs to call intrinsics. @DaniPopes did `enum AA` in your example has `Debug` implementation?

**Comment by Nikaidou-Shinku** [comment]
I encountered the same problem, in my case `const FOO: [&'static str; 1] = ["bar"]` would trigger the same problem.

**Comment by Nilstrieb** [comment]
I'm taking a look at fixing this by handling the new intrinsics correctly.

**Comment by Nilstrieb** [comment]
Uh, r-a's implementation of `panic_fmt` calls `std::fmt::format`... I guess that then in turn panics or something like that. I think it needs to call `const_panic_fmt`, just like rustc. #15222 introduced this with no information whatsoever :thinking: @HKalbasi 
I suspect what happened is that something now cause `std::fmt::format` to panic, leading to pathological escpaing.
Because all the code is doing is escaping the panic message, and then escaping that, and escaping that etc, in just a few seconds we get gigabytes of backslashes, lmao.

**Comment by DrewMcArthur** [comment]
@Nilstrieb in case it's helpful at all, i added this comment to a closed issue before i realized that this and #16934 were both still open.  memory is leaking for me whenever vs code tries to format-on-save, which sounds like it's related to your last comment here.

https://github.com/rust-lang/rust-analyzer/issues/16140#issuecomment-2016859383
